### PR TITLE
Addressing #5159 (IPv6 dask-worker support) by propagating empty host…

### LINKED ISSUE #5159: `dask-worker` broken with IPv6?
**What happened**: I was trying to figure out why tasks weren't completing on our dask cluster. Clicking the "Logs" button for an IPv6 worker on the scheduler dashboard (`/info/main/workers.html`) hung. Clicking "logs" for the IPv4 workers did not hang. Worked backward to this MWE, which I'm guessing is the cause. Not 100% sure though.

**What you expected to happen**: The "Logs" button should work at least. Ideally, fixing this will also allow the tasks that weren't completing to complete, but am unsure of that aspect. I am currently testing if this will also fix the hung jobs by disabling our IPv6 workers; will update in awhile.

**Minimal Complete Verifiable Example**:

On a machine accessible only over IPv6, running:

```sh
$ dask-worker --listen-address tcp://:8811 <scheduler>
```

And then

```sh
$ lsof -i :8811
```

Clearly shows that the worker is listening only on IPv4, not IPv6.

**Anything else we need to know?**:

**Environment**:

- Dask version: 2021.7.2
- Python version: 3.8
- Operating System: Ubuntu in Docker
- Install method (conda, pip, source): pip

### PR DESCRIPTION
**Pull request by wwoods**
… correctly

- [x] Closes #5159. Seems to support dask-workers listening on both IPv4 and IPv6, as well as either/or. Solves dashboard log access problem mentioned in that bug, and appears to allow work to process.
- [x] Tests added / passed -- tests have been added. Related bug mentioned in comments split into #5206. This PR should be good to go if the CI passes.
- [x] Passes `black distributed` / `flake8 distributed` / `isort distributed`

@jacobtomlinson Let me know what you think. Thanks!

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by jrbourbeau** [comment]
add to allowlist

**Comment by wwoods** [comment]
Given tests are failing, but they're also failing for the commit that this is based on, I will probably rebase this PR to https://github.com/dask/distributed/commit/9569e08f28df3b870048ce7ae8b44961973af6c1 so we get a better idea of which failures are actually a result of this change.

**Comment by wwoods** [comment]
Quick note -- I found a bug in the tests that could reasonably crop up in production while I was debugging some of the other test issues.

The other bug is that, prior to this PR, workers would _only_ fetch task data from a local worker when one is available, even if that worker is busy. One could imagine a set of long-running tasks that tie up the local worker, while a remote worker also has the data available and is not busy. In this case, the worker would wait on the local, long-running tasks, rather than just asking the remote worker for the data.

There's a TODO on the fix -- I didn't see any `time.monotonic` usage in `distributed/worker.py`, so I was presuming that sort of timeout was to be avoided. The new `busy_workers` attribute will reset its status for workers which hold task data when all such workers were previously marked busy; however, this is not ideal as a local worker could be rejected for quite some time before being the only option with the data.

So, I was thinking of resetting the `busy_workers` attribute when a task dependency is newly requested. I'm guessing that would be a good approach?

2nd TODO is line 2187 in distributed/workers.py should use `workers_not_busy`. Wanted to get this changeset pushed earlier though since I think it will fix the remaining broken tests, leaving only this parallel issue to be resolved.

File: distributed/worker.py
Comment (jacobtomlinson) [inline distributed/worker.py:562]: I've not seen `assert` used like this before. Would it be better to raise an exception here more explicitly?

Comment (wwoods) [reply distributed/worker.py:562]: It would be different but same effect -- happy to make the change if you would like.

Other than this, does the PR look good?

Comment (jacobtomlinson) [reply distributed/worker.py:562]: > It would be different but same effect

Functionally it would be the same but by using a more familiar pattern the code is more maintainable. (I'm sure @mrocklin wrote a good blog post about the value of writing code that a novice contributor could understand but I am failing to find it)

Otherwise yeah this seems like a great change.

Comment (wwoods) [reply distributed/worker.py]: Sure, change to `ValueError` pushed.

**Review by jacobtomlinson — APPROVED** [review]

**Comment by wwoods** [comment]
Just in case something about this PR is waiting on the CI tests to pass -- the remaining failures appear to be transient failures, unrelated to these changes.

**Comment by jacobtomlinson** [comment]
Yeah sorry I was waiting for CI to finish before merging and forgot to come back here.

**Comment by jacobtomlinson** [comment]
Thanks @wwoods

**Pull request merged**
